chalk-slg/src/logic.rs

```diff
@@ -103,6 +103,25 @@ impl<C: Context> Forest<C> {
         }
     }
 
+    pub(super) fn may_invalidate(
+        &mut self,
+        table: TableIndex,
+        subst: &C::Substitution,
+    ) -> bool {
+        let mut strands = self.tables[table].strands_mut();
+
+        // If we don't have any strands at all, then we are
+        // intentionally conservative and say it may_invalidate.
```

Comment:
Say more about this? It surprises me a bit. Well, I guess it doesn't hurt -- they will request the next answer then, and get back nothing -- but I'm not sure why it's worth making a special case for it.

Reply:
I'm not yet entirely sure why, but without this check, there were two tests that failed, `solve::test::prove_infer` and `solve::test::suggested_subst`, which appears to be because they expected to get `Unknown` but were instead giving `subst`. I think the other option is a change in `make_solution`, but I'm not positive what that looks like.

Reply:
curious

Reply:
OK, I did a bit of investigating. I found the problem. The problem is that our main loop invokes `peek_answer` and then `may_invalidate` -- but if `peek_answer` gets an answer already, then there is no strand left for `may_invalidate` to examine. So it is correct that no strands will invalidate, but incorrect that no *future answers* may do so. 

To fix this, we should pass in the current answer index to `may_invalidate`. This is also needed -- now that I think about -- for the caching scenario I expect in the future, where the answers have already been computed as part of some other query and we are just replaying them.